lang/LangPrimSource/SC_AudioDevicePrim.cpp

```diff
@@ -125,6 +131,56 @@ int listDevices(struct VMGlobals* g, int type) {
     return 1;
 }
 
+#elif defined(SC_AUDIO_API_PORTAUDIO)
+int listDevices(struct VMGlobals* g, int type) {
+    int numDevices;
+    PaError err = Pa_Initialize();
+    if (err != paNoError)
+        return 0;
+
+    numDevices = Pa_GetDeviceCount();
+    if (numDevices < 0)
+        return 0;
+
+    std::vector<const PaDeviceInfo*> deviceInfos;
```

Comment:
@brianlheim do I need to deallocate this vector?

Reply:
nope, that is the magic of [RAII](https://en.cppreference.com/w/cpp/language/raii) -- in short, much of modern C++ relies on the concept of treating an object's constructor as an initialization and resource acquisition step, and treating its destructor (here, called when we exit the function) as a resource freeing step. opening files (std::fstream), allocating memory (std::vector and other std containers), creating threads (std::thread), and acquiring/releasing a lock (std::lock_guard) all can work this way in C++.